Added image management command to purge and regenerate renditions

Hitansh-Shah opened the pull request:
I have added an image management command `wagtail_update_image_renditions` in images app of wagtail for regenerating and purging image renditions. I have included the support for additional parameters like `--purge` and `purge-only` and I have tried to keep the overall implementation simple. This is my first pull request so pardon me if I have made any silly mistakes along the way 😅. I ran the tests and I didn't find any errors. I have not yet written tests for this code but I will once the code is checked and finalised. Please review it @lb- . Thanks

Fixes #8107

lb- commented:
If merged, should credit previous work https://github.com/wagtail/wagtail/pull/759
Also, could supersede https://github.com/wagtail/wagtail/pull/6595 (credit for that would be good also)

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,50 @@
+        parser.add_argument(
+            '--purge-only',
+            action='store_true',
+            help='Only purge all image renditions'
```
```suggestion
            help='Purge all image renditions without regenerating them'
```

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,50 @@
+                    rendition.image.get_rendition(rendition.filter)
+                except:
+                    self.stdout.write(self.style.ERROR('Could not regenerate rendition for %s', rendition.image.title))
+            self.stdout.write(self.style.SUCCESS('Succesfully regenerated image renditions!'))
```
might be nice to just run the success message once at the end, ideally with a count of the renditions regenerated.

Something like
```
self.stdout.write(self.style.SUCCESS(f'Successfully regenerated {total} image renditions'))
```

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,50 @@
+                try:
+                    rendition.image.get_rendition(rendition.filter)
+                except:
+                    self.stdout.write(self.style.ERROR('Could not regenerate rendition for %s', rendition.image.title))
```
is it possible to log the error to the stdout also?

Hitansh-Shah replied:
Not sure myself, should I use stderr to stay on safe side?

lb- replied:
sorry, my message was confusing.

What I mean is that is it possible to also output some/all of the error thrown, but after looking at some other commands that do something similar, this is probably not needed.

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,50 @@
+from django.core.management.base import BaseCommand, CommandError
+from wagtail.images.models import Rendition, Image
```
We will need to likely use the `get_image_model`  so this code can support custom image models

https://docs.wagtail.org/en/stable/advanced_topics/images/custom_image_model.html#module-wagtail.images

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,50 @@
+
+
+    def handle(self, *args, **options):
+        renditions = Rendition.objects.all()
```
This code will need to get the rendition model via the class method `get_rendition_model` on the image model that is resolved from `get_image_model`

This is undocumented but see the file
wagtail/images/models.py -> https://github.com/wagtail/wagtail/blob/2.16/wagtail/images/models.py#L299

lb- reviewed (changes requested):
@Hitansh-Shah - thanks so much for taking the time to contribute.

This is looking like a great start, I have a bunch of inline comments below, please feel free to refine the suggested changes as suited but just trying to reword things a bit.

Once you have worked through these changes, I think this is good to start on unit tests & documentation next.

So next steps would be
1. Check the CI failures (probably linting) and fix up
2. Documentation (good to write this before unit tests, gives you a sense of how to think about the command being used by others, also might help with some wording refinements for the help texts)
3. Please also add a reference to this new command in `docs/topics/images.rst` (see the 'Changing globally`' section has an example code for removing all renditions).
4. Unit tests (including a test for a custom image model / rendition model - there should be a model in the existing tests you can use)
5. It is ok if this does not get done, but it might be nice to add a `--dryrun` option to the command that does not delete anything but logs what would have been deleted/re-created. This can always be a future enhancement so no pressure.

Thanks again, really amazing stuff.

Hitansh-Shah commented:
Thanks @lb- for reviewing. I will update the code as you suggested and will try to implement the next steps you mentioned. I really appreciate the guidance.

Hitansh-Shah commented:
@lb- I have made the changes you suggested along with some other improvements.

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,71 @@
+        )
+
+    def handle(self, *args, **options):
+        renditions = get_image_model().get_rendition_model().objects.all()
```
It would be good to return early here if there are no renditions, maybe with a message

`self.stdout.write('No image renditions found')`

lb- commented:
Thanks, feel free to ping me on this when ready for another review.

Hitansh-Shah commented:
@lb- I have made 3 commits for:
1)Early return if there are no renditions.
2)Adding the command to the docs.
3)Adding the reference of the command to `images.rst`

Please review the changes and tell me if any improvements or changes are to be made. In the meantime I will work on the tests 🚀

Hitansh-Shah commented:
@lb- Can you guide me a little about writing tests? I am kinda confused. So there is `test_management_commands.py` file in `/wagtail/tests`. So should I include the test in this file or should I create a new similar file in `/wagtail/images/tests`? Also if you could guide me a little about what should I include in the test, that would be great. I don't have much experience in writing tests but I will get a hang of it eventually as I write more and more tests. Thanks for the help 😇.

lb- commented:
@Hitansh-Shah here is a rough approach

Create a new file `wagtail/images/tests/test_management_commands.py`

Here is a rough starting test (note, may not work but should get you enough errors to get it running)
```
from io import StringIO

from django.core import management
from django.test import TestCase

from wagtail.images import get_image_model


class TestUpdateImageRenditions(TestCase):

    # this line is critical 
    fixtures = ["test.json"]

    def run_command(self):
        output = StringIO()
        management.call_command("wagtail_update_image_renditions", stdout=StringIO())
        output.seek(0)

        return output

    def test_image_renditions(self):
        # use get_image_model
        # first check there are X renditions (likely none as not any in fixtures)
        # run the command
        output = self.run_command()
        # check the message says something about zero things
        output_string = output.read()
        self.assertIn("Zero something", output_string)

```
Code is cherry picked from existing management command tests `wagtail/tests/test_management_commands.py`

Hitansh-Shah commented:
@lb- I have added test for the `wagtail_update_image_renditions` management command. Please review it and tell me if any changes or enhancements are to be made. 
I am little nervous about the `rendition` model that I have added in `test.json`, check and tell me if that's the correct way to do it. Also all tests that I have added seem to work fine except `self.test_image_renditions_with_purge`. It is somehow failing because the command throws an error and therefore the number of  successful renditions regenerations are not matching. I am attaching the error message, please have a look at it and tell me where have I messed up 😅. Thanks.
![image](https://user-images.githubusercontent.com/76250591/159659461-efeea874-d392-4378-9f8d-d020df39534f.png)
I updated the command definition so that it gives me the cause of error, I am attaching the error message too.
![image](https://user-images.githubusercontent.com/76250591/159660117-cff1dabc-211e-44e9-9b21-c01b12e4aa7f.png)

Hitansh-Shah commented:
@lb- As you mentioned earlier, I have tried to implement a very simple/basic approach for `--dryrun`. Please have a look at it and tell me what you think about it.

lb- commented on docs/reference/management_commands.rst:
```diff
@@ -125,3 +125,26 @@ search_garbage_collect
+This command provides the ability to regenerate image renditions.
+This is useful if you have deployed to a server where the image renditions have not yet been generated or you have changed the underlying image rendition behaviour and need to ensure all renditions are created again.
+
+This does not remove rendition images that are unused, this can be done by clearing the folder using ``rm -rf`` or similar, once this is done you can then use the management command to generate the renditions.
```
I added these notes for clarity.

Hitansh-Shah replied:
Got it. Thanks!

lb- reviewed (changes requested):
Hey @Hitansh-Shah  this is looking good, a few more things to fix up.

Firstly, this does not seem to regenerate rendition images if they have been removed manually (e.g. via `rm -rf`) instead is just says 'no images found'. Could you please look into this.

I have done the following on your remote branch
* rebased on `main`, squashed and cleaned up commits
* Removed the `--dry-run` code, let's solve that in the future as we do not want the scope to creep on this, otherwise it may never get in
* I have backed up your original work here https://github.com/lb-/wagtail/tree/backup/hintansh-image-management-command
* I have fixed a few typos, checked the docs build and checked all the commands. they all work great except for the use case described above
* I have also removed the extensive `!` usage in the command feedback if that is ok and updated the help text slightly.
* Hopefully now we can see if the build works in the CI and you can resolve any tests that are not working.

Hitansh-Shah commented:
> Firstly, this does not seem to regenerate rendition images if they have been removed manually (e.g. via `rm -rf`) instead is just says 'no images found'. Could you please look into this.

Oh yeah that is correct. However to generate renditions I have used `get_renditions()` which accepts a required parameter `filter` which can be retrieved from existing rendition object, so is there a way to know the `filter` parameter if the rendition is deleted?

lb- commented on docs/reference/management_commands.rst:
```diff
@@ -125,3 +125,26 @@ search_garbage_collect
 
 Wagtail keeps a log of search queries that are popular on your website. On high traffic websites, this log may get big and you may want to clean out old search queries. This command cleans out all search query logs that are more than one week old (or a number of days configurable through the :ref:`WAGTAILSEARCH_HITS_MAX_AGE <wagtailsearch_hits_max_age>` setting).
+
+
```
Probably only need one new line here
```suggestion
```

lb- commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,75 @@
+        else:
+            for rendition in renditions:
+                try:
+                    rendition.image.get_rendition(rendition.filter)
```
For some reason this is not working as expected when the rendition file is removed from the filesystem.

Hitansh-Shah replied:
I guess I figured out what exactly is happening here. When we delete the rendition image file from filesystem (for eg: deleting from /bakerydemo/media/images in bakerydemo example) the file is deleted but the rendition model with that image's data is still stored somewhere. Moreover when the `get_rendition` method is called with same `filter` the rendition is not generated again (because the original image and filter are same) and therefore the rendition image is not reflected in images folder.

Hitansh-Shah replied:
One solution to this can be we use the code of `--purge` option by default. Therefore the command will first delete the existing rendition and then generate it.

lb- replied:
haha classic, we had the same idea, go for it.

Thanks for working through this, I know it is not as simple as we first thought.

This may also make your tests simpler so hopefully that other issue does not keep happening.

Hitansh-Shah replied:
😄 I will update the command and tests and push soon. Hoping to not break anything 😬

lb- commented:
> > Firstly, this does not seem to regenerate rendition images if they have been removed manually (e.g. via `rm -rf`) instead is just says 'no images found'. Could you please look into this.
> 
> Oh yeah that is correct. However to generate renditions I have used `get_renditions()` which accepts a required parameter `filter` which can be retrieved from existing rendition object, so is there a way to know the `filter` parameter if the rendition is deleted?

As I understand it the rendition is a DB row that refers to an image on the filesystem, so deleting the image in the filesystem should not delete the rendition entry in the DB and it is the rendition that contains its filter, not the image itself.

Hitansh-Shah commented:
Hey @lb- , I have done the changes you suggested. However one test is still failing (for purge and generate which is now default). It is saying that " No such file or directory: `wagtail/wagtail/tests/test-media/original_images/missing.jpg`". I am attaching the error screenshot.
![image](https://user-images.githubusercontent.com/76250591/161815288-da4fed80-c755-4051-ba36-dfdabf87966c.png)

lb- commented:
@Hitansh-Shah please push up your changes, even if they may be failing, it makes it easier to help you debug.

lb- commented:
> missing.jpg

I think you may need to revise your tests, remove the addition of the rendition in the test.json and maybe remove the fixtures completely from being used in your test.

Instead 'create' the renditions inside the test either in the test class setup method `def setUp(self):` or as a test method, this includes creating the image.

See how `TestRenditions` does it

Hitansh-Shah commented:
Thanks @lb- . I have updated some code. Please have a look and tell me what you think.

zerolab commented on wagtail/images/management/commands/wagtail_update_image_renditions.py:
```diff
@@ -0,0 +1,56 @@
+                    )
+            self.stdout.write(
+                self.style.SUCCESS(
+                    f"Successfully purged {success_count} image renditions"
```
nitpick: if only one rendition is purged the "Successfully purged 1 image renditions" is incorrect

Hitansh-Shah replied:
Oh yeah that slipped through my mind 😬. We can either replace `renditions` with `rendition(s)` or I can put some `if else` nesting

lb- reviewed (approved):
@Hitansh-Shah thanks for this, I have updated the docs and also revised the messaging to be suitable for 1 or more rendition changes.

Just a simple approach `...image rendition(s)\n"`

FYI @zerolab

lb- commented:
@Hitansh-Shah - thank you!

Also - a thank you to @onno-timmerman @damianmoore who worked on earlier versions of this solutions many years ago on which this implementation was based.

enzedonline commented:
Would appreciate this being added to the docs - spent an age working out how to do this after getting post v6 upgrade image errors, then find this pr from a google search ...
https://docs.wagtail.org/en/stable/advanced_topics/images/renditions.html would be a good place

lb- commented:
@enzedonline  this is added to the docs (2 places, 3 if you count the release notes) but I can see how it may be confusing that we did not add it to the renditions page.

See https://docs.wagtail.org/en/stable/topics/images.html#changing-globally & https://docs.wagtail.org/en/stable/reference/management_commands.html#wagtail-update-image-renditions & 5.2 release notes (with a docs link) https://docs.wagtail.org/en/stable/releases/5.2.html#other-features

In general if you want to know what management commands exist, we do catalogue them all on the management commands page.

Anyway, I have a PR to add it on that page also https://github.com/wagtail/wagtail/pull/11787

enzedonline commented:
> @enzedonline this is added to the docs (2 places, 3 if you count the release notes) but I can see how it may be confusing that we did not add it to the renditions page.

Thanks for this - I'd combed through the Advanced Topics > Image section & the Reference sections for Image and Rendition. Hadn't really thought of it as a topic for using in templates.

The struggle for me was that I needed to purge the renditions for the problem images, but the renditions method on the Image object requires a filter rather than just returning all the renditions for that image. Ended up using the management command to blitzing all the renditions for all images.
